### PR TITLE
Type `add_message` and add `MessageLocationTuple`

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

This adds typing to `add_message` and adds a new `NamedTuple` to `typing.py`. I add a deprecation warning for `Message`, but I'm not sure if it is even necessary. Is `Message` part of the public API? If not I can remove it.

The `type: ignore` in `tests/testutils/test_output_line.py` will be fixed in a follow-up PR. `OutputLine` needs a little refactor to only accept `int` for `column` and `line`. This will bring it in line with all other code that references `column` or `line`, where those are also `int`.

### CODE REVIEW

File: pylint/message/message_handler_mix_in.py [hunk 1]
```diff
@@ -267,14 +279,20 @@ def check_message_definition(message_definition, line, node):
                         f"Message {message_definition.msgid} must provide Node, got None"
                     )
 
-    def add_one_message(
-        self, message_definition, line, node, args, confidence, col_offset
-    ):
+    def add_one_message(  # type: ignore # MessagesHandlesMixIn is always mixed with PyLinter
```
Comment (Pierre-Sassoulas) [inline pylint/message/message_handler_mix_in.py:282]:
```suggestion
    def add_one_message(  # type: ignore # MessagesHandlerMixIn is always mixed with PyLinter
```

File: pylint/message/message.py [hunk 2]
```diff
@@ -28,7 +32,19 @@
...
+        cls,
+        msg_id: str,
+        symbol: str,
+        location: Union[Tuple[str, str, str, str, int, int], MessageLocationTuple],
+        msg: str,
+        confidence: Optional[Confidence],
+    ):
+        if isinstance(location, tuple):
+            warn(
+                "In pylint 3.0, Messages will only accept a MessageLocationTuple as location parameter",
```
Comment (Pierre-Sassoulas) [inline pylint/message/message.py:45]: :+1:

File: pylint/message/message_handler_mix_in.py [hunk 3]
```diff
@@ -226,9 +232,15 @@ def is_one_message_enabled(self, msgid, line):
                 return self._msgs_state.get(msgid, fallback)
             return self._msgs_state.get(msgid, True)
 
-    def add_message(
-        self, msgid, line=None, node=None, args=None, confidence=None, col_offset=None
-    ):
+    def add_message(  # type: ignore # MessagesHandlesMixIn is always mixed with PyLinter
+        self: "PyLinter",
+        msgid: str,
```
Comment (Pierre-Sassoulas) [inline pylint/message/message_handler_mix_in.py:237]: `msgid` is a special string with one letter between C W I F R, then 4 digits (C0111)  and `symbol` is a mix of alphabetic character and `-` (missing-docstring). A lot of function in `pylint.message` can takes both but not everything is named correctly right now. You can check `pylint.message.MessageIdStore` & co.


```suggestion
        msgid_or_symbol: MessageId, # MessageId already exists, it seems it would require a refactor or a new type
```

**Review by Pierre-Sassoulas — COMMENTED** [review]
I like this change. But maybe the msgid/symbol typing is something that should be done prior to this one ?

File: pylint/message/message_handler_mix_in.py (hunk 3 above)
Comment (DanielNoord) [reply pylint/message/message_handler_mix_in.py:237]: You mean create a new type alias called `MessageId` with the specifics discussed in your comment?
I can try and do this, but I think this might create some issues by being too specific. As in, `mypy` will then want us to make sure that every `msgid_or_symbol` conforms to the rules of `msgid_or_symbol` and we might need to add checks all over the code base for this. Still, this is something to try!

Personally I would like to do the following sequence: this PR > `OutputLine` refactor + typing > `msgid_or_symbol` refactor. This will give me the typing of the first two PRs wil working on the `msgid_or_symbol` refactor.

File: pylint/message/message_handler_mix_in.py [hunk 4]
```diff
@@ -226,9 +232,15 @@ def is_one_message_enabled(self, msgid, line):
                 return self._msgs_state.get(msgid, fallback)
             return self._msgs_state.get(msgid, True)
 
-    def add_message(
-        self, msgid, line=None, node=None, args=None, confidence=None, col_offset=None
-    ):
+    def add_message(  # type: ignore # MessagesHandlesMixIn is always mixed with PyLinter
```
Comment (DanielNoord) [inline pylint/message/message_handler_mix_in.py:235]:
```suggestion
    def add_message(  # type: ignore # MessagesHandlerMixIn is always mixed with PyLinter
```

File: pylint/message/message_handler_mix_in.py (hunk 3 above)
Comment (Pierre-Sassoulas) [reply pylint/message/message_handler_mix_in.py:237]: I don't know if mypy can check that a string type is a subset of str. I was thinking of `MessageId` as simply a semantic alias for `str` (The `MessageId` class is both str value (C0111 / missing-docstring) retrieved from the MessageStore using only one of the value and guaranteed to be coherent).

File: pylint/message/message_handler_mix_in.py [hunk 5]
```diff
@@ -267,14 +279,20 @@ def check_message_definition(message_definition, line, node):
...
 
-    def add_one_message(
-        self, message_definition, line, node, args, confidence, col_offset
-    ):
+    def add_one_message(  # type: ignore # MessagesHandlerMixIn is always mixed with PyLinter
+        self: "PyLinter",
+        message_definition: "MessageDefinition",
+        line: Optional[int],
+        node: Optional[nodes.NodeNG],
+        args: Union[str, Tuple[Union[str, int], ...], None],
```
Comment (cdce8p) [inline pylint/message/message_handler_mix_in.py:287]: Is it really limited to only these three types? This is the signature for `str.format`:
```py
def format(self, *args: object, **kwargs: object) -> str: ...
```

File: pylint/typing.py [hunk 6]
```diff
@@ -50,3 +50,13 @@ class ErrorDescriptionDict(TypedDict):
 CheckerStats = Dict[
     str, Union[int, "Counter[str]", List, Dict[str, Union[int, str, Dict[str, int]]]]
 ]
+
+
+# Tuple with information about the location of a to-be-displayed message
```
Comment (cdce8p) [inline pylint/typing.py:55]: Why not use a docstring for it?

File: tests/message/unittest_message.py [hunk 7]
```diff
@@ -1,29 +1,31 @@
...
-                location_value["column"],
-            ],
+            location=(
+                location_value.abspath,
+                location_value.path,
+                location_value.module,
+                location_value.obj,
+                location_value.line,
+                location_value.column,
+            ),
```
Comment (cdce8p) [inline tests/message/unittest_message.py:26]: Why is this necessary? Wasn't the whole point so that you could do `location=location_value`?

File: pylint/message/message_handler_mix_in.py (hunk 5 above)
Comment (DanielNoord) [reply pylint/message/message_handler_mix_in.py:287]: Yes, I tested this and `mypy` complains whenever a call has other types for `args`. So I think we cans safely assume this is correct.

File: tests/message/unittest_message.py (hunk 7 above)
Comment (DanielNoord) [reply tests/message/unittest_message.py:26]: Sometimes after your reviews I wonder if I was even thinking when submitting a PR... 😅

Comment (Pierre-Sassoulas) [reply tests/message/unittest_message.py]: I missed this one, damn ! 😄

File: pylint/message/message_handler_mix_in.py (hunk 3 above)
Comment (cdce8p) [reply pylint/message/message_handler_mix_in.py:237]: > Not sure if this would actually provide a benefit.

I agree. There are other changes which are much more useful.
So probably nothing we should pursue in the immediate future.

File: pylint/message/message_handler_mix_in.py (hunk 5 above)
Comment (cdce8p) [reply pylint/message/message_handler_mix_in.py:287]: You just need to replace all reference, then it should work.
* `BaseChecker.add_message`
* `MessagesHandlerMixin.add_message`
* `MessagesHandlerMixin.add_one_message`

I would also recommend to use `args: Any` for now. We can always narrow it down later, if we choose to.

File: tests/testutils/test_output_line.py [hunk 8]
```diff
@@ -19,7 +19,7 @@ def inner(confidence: Confidence = HIGH) -> Message:
         return Message(
             symbol="missing-docstring",
             msg_id="C0123",
-            location=[
+            location=[  # type: ignore
```
Comment (cdce8p) [inline tests/testutils/test_output_line.py:22]: This should probably be a `MessageLocationTuple`. Unfortunately, you'll still need `type: ignore` comments for `"line"` and `"column".

Comment (DanielNoord) [reply tests/testutils/test_output_line.py:22]: Yes, but I want to change this with an upcoming refactor/typing of `OutputLine`. See the PR description. I can also do that first if you want me to, but I think it might be better to do this first and then do `OutputLine` and its tests.

Comment (cdce8p) [reply tests/testutils/test_output_line.py:22]: Sorry. Probably shouldn't review PRs that late 😅

Comment (DanielNoord) [reply tests/testutils/test_output_line.py:22]: No worries! As you know it is truly appreciated!

File: pylint/message/message_handler_mix_in.py (hunk 5 above)
Comment (DanielNoord) [reply pylint/message/message_handler_mix_in.py:287]: I don't really understand the first half of your comment. Doesn't this PR already add the `args` typing to all those three places?

I think we can add this typing. `mypy` checks all calls to any of the three functions and checks whether `args` has the correct typing. This was the reason I added `int` to the `Tuple`, `mypy` notified me of that possibility. Perhaps `mypy` is not doing what I think it is, but I believe there is no way for `args` to be anything other than what is now specified in the typing without `mypy` telling us about it.

Comment (cdce8p) [reply pylint/message/message_handler_mix_in.py:287]: That would be the case if the codebase was fully type. However `mypy` doesn't check the body of untyped function definitions. So there is the likely case that `Union[str, Tuple[Union[str, int], ...], None]` doesn't cover it.

Additionally, this only covers the our code but what about plugins? Can they for example pass a `float` with `args`.

Thus, I would recommend to replace it with `Any` (at the very least for now).

Comment (DanielNoord) [reply pylint/message/message_handler_mix_in.py:287]: I think #4940 made it so all function scopes which contain an `add_message` are typed. But you are right, there is a real possibility that plugins pass other types of `args`. I will change the typing to `Any`.

Comment (cdce8p) [reply pylint/message/message_handler_mix_in.py]: Sadly, we aren't quite there yet. Adding types there also won't be that easy as it was for #4940. This one here took me < 1min to find:
https://github.com/PyCQA/pylint/blob/b3336c77378f88eb1428d6ff5b6e78bba1177f40/pylint/checkers/base.py#L689-L691

File: pylint/checkers/base_checker.py [hunk 9]
```diff
@@ -109,10 +111,14 @@ def get_full_documentation(self, msgs, options, reports, doc=None, module=None):
...
     def add_message(
-        self, msgid, line=None, node=None, args=None, confidence=None, col_offset=None
-    ):
-        if not confidence:
-            confidence = UNDEFINED
+        self,
+        msgid: str,
+        line: Optional[int] = None,
+        node: Optional[nodes.NodeNG] = None,
+        args: Optional[Any] = None,
```
Comment (cdce8p) [inline pylint/checkers/base_checker.py:118]: `Any` also covers `None`

Comment (DanielNoord) [reply pylint/checkers/base_checker.py:118]: Ah, I did this because of this [mypy flag](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-no-implicit-optional). It didn't mention `Any`. My bad!

**Review by cdce8p — APPROVED** [review]
Looks good to me. @Pierre-Sassoulas Would you mind taking another look?

File: pylint/message/message_handler_mix_in.py (hunk 3 above)
Comment (Pierre-Sassoulas) [reply pylint/message/message_handler_mix_in.py:237]: Yes the simple change to do is to keep the string type and rename args ``msgid`` to ``msgid_or_symbol`` (the convention I used in ``pylint.messages``)

File: pylint/checkers/base_checker.py [hunk 10]
```diff
@@ -109,10 +111,14 @@ def get_full_documentation(self, msgs, options, reports, doc=None, module=None):
         return result
 
     def add_message(
-        self, msgid, line=None, node=None, args=None, confidence=None, col_offset=None
-    ):
-        if not confidence:
-            confidence = UNDEFINED
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/base_checker.py:115]: I don't see where the default confidence is set now ? Or did we remove the UNDEFINED value in favor of None ?

File: pylint/message/message.py [hunk 11]
```diff
@@ -28,7 +32,42 @@
...
+        location: Union[Tuple[str, str, str, str, int, int], MessageLocationTuple],
+        msg: str,
+        confidence: Optional[Confidence],
+    ) -> "Message":
+        if not isinstance(location, MessageLocationTuple):
+            warn(
+                "In pylint 3.0, Messages will only accept a MessageLocationTuple as location parameter",
+                DeprecationWarning,
+            )
         return _MsgBase.__new__(
```
Comment (Pierre-Sassoulas) [inline pylint/message/message.py:71]:
```suggestion
        confidence = UNDEFINED if confidence is None else confidence
        return _MsgBase.__new__(
```

**Review by Pierre-Sassoulas — APPROVED** [review]
Thanks, you're almost single-handedly typing pylint :)

File: pylint/message/message_handler_mix_in.py (hunk 3 above)
Comment (DanielNoord) [reply pylint/message/message_handler_mix_in.py:237]: Couldn't that create problems with extensions explicitly using `msgid`?

File: pylint/checkers/base_checker.py (hunk 10 above)
Comment (DanielNoord) [reply pylint/checkers/base_checker.py:115]: On L116 this functions calls `self.linter.add_message(msgid, line, node, args, confidence, col_offset)` which sets the default confidence here:
https://github.com/PyCQA/pylint/blob/b3336c77378f88eb1428d6ff5b6e78bba1177f40/pylint/message/message_handler_mix_in.py#L229-L246

I thought this was duplicate code.

File: pylint/message/message_handler_mix_in.py (hunk 3 above)
Comment (Pierre-Sassoulas) [reply pylint/message/message_handler_mix_in.py:237]: Yeap, probably a breaking change, let's put in 3.0 or not at all.

File: pylint/checkers/base_checker.py (hunk 10 above)
Comment (cdce8p) [reply pylint/checkers/base_checker.py:115]: Indeed, looks like `lister.add_message` handles that for us already.

File: pylint/message/message.py (hunk 11 above)
Comment (cdce8p) [reply pylint/message/message.py:71]: Why set it in the NamedTuple if it's handled by `add_message` already? Am I missing something?

File: pylint/message/message_handler_mix_in.py (hunk 3 above)
Comment (cdce8p) [reply pylint/message/message_handler_mix_in.py:237]: IMO probably not worth doing. It could break too many things.

File: pylint/message/message.py (hunk 11 above)
Comment (DanielNoord) [reply pylint/message/message.py:71]: I think this was suggested because Pierre thought I had removed it erroneously.
